### ISSUE TITLE
Country-specific Address Formats

### ISSUE THREAD

**Issue by pyrog** [osm-search/Nominatim#213]
I would like to test existing [Country Address Format](http://wiki.openstreetmap.org/wiki/Nominatim/Country_Address_Format). The goal is to create some for my country.

Could you give me some query/examples that display adresses in theses format ?

**Comment by freyfogle** [comment]
Just to update this issue, the address-formatting repo has moved to https://github.com/OpenCageData/address-formatting 

we now have 100% coverage and 86% of territories have at least one test. Slowly I am adding more, I expect 100% test coverage by 1 April 2016. Of course there are many, many edge cases, so more eyeballs and tests are definitely needed. All help much appreciated

**Comment by RafalR** [comment]
when addresses will be fixed?
Example 1:
```
all valid addresses -> http://overpass-turbo.eu/s/hVV
adama mickiewicza 6, lublin -> incorrect! first result from zamość,
should be http://www.openstreetmap.org/way/319339020
adama mickiewicza 8, lublin -> incorrect! first result from zamość,
should be http://www.openstreetmap.org/node/2604114620
adama mickiewicza 10, lublin -> incorrect! first result from zamość,
should be http://www.openstreetmap.org/node/2604092061
adama mickiewicza 12, lublin -> incorrect! first result from zamość,
should be http://www.openstreetmap.org/way/352817713
adama mickiewicza 14, lublin -> incorrect! first result from zamość,
should be http://www.openstreetmap.org/node/2604099139
adama mickiewicza 16, lublin -> incorrect! first result from chełm,
should be http://www.openstreetmap.org/way/352817730
adama mickiewicza 18, lublin -> incorrect! first result from zamość,
should be http://www.openstreetmap.org/way/352817758
adama mickiewicza 19, lublin -> correct :)
adama mickiewicza 21, lublin -> correct :)
adama mickiewicza 23, lublin -> correct :)
23...107 -> correct :)
adama mickiewicza 107, lublin -> correct :)
```
Example 2:
```
all valid addresses -> http://overpass-turbo.eu/s/hVY
tadeusza kościuszki 1, lublin -> incorrect! first result from włodawa,
should be http://www.openstreetmap.org/node/833340653
tadeusza kościuszki 2, lublin -> correct :)
tadeusza kościuszki 3, lublin -> incorrect! first result from puławy,
should be http://www.openstreetmap.org/node/833340651
tadeusza kościuszki 4, lublin -> incorrect! first result from puławy,
should be http://www.openstreetmap.org/node/814991082
tadeusza kościuszki 5, lublin -> incorrect! first result from puławy,
should be http://www.openstreetmap.org/node/833340647
tadeusza kościuszki 6, lublin -> incorrect! first result from puławy,
should be http://www.openstreetmap.org/way/67508881
tadeusza kościuszki 7, lublin -> incorrect! first result from puławy,
should be http://www.openstreetmap.org/node/833340681
tadeusza kościuszki 8, lublin -> incorrect! first result from puławy,
should be http://www.openstreetmap.org/node/814991080
tadeusza kościuszki 10, lublin -> incorrect! first result from puławy,
should be http://www.openstreetmap.org/node/814991077


```

**Comment by Bibi56** [comment]
> Just to update this issue, the address-formatting repo has moved to https://github.com/OpenCageData/address-formatting

@lonvia, Apparently one implementation of this exists in PHP: https://github.com/predicthq/address-formatter-php, any good reason (except time ;-)) not to use it?

**Comment by lonvia** [comment]
This should be much easier to implement now that we've moved the computation of the display name to Python. Here is roughly what I'm thinking how this could work:

* add a new parameter to the API endpoints that return a `display_name` (search, reverse, lookup) that allows to select the format
* pull computation of the display_name out of the [localize function](https://github.com/osm-search/Nominatim/blob/00b3ace3cf92978253557290b0ef12e85d05d7bf/src/nominatim_api/results.py#L235) and call instead [here](https://github.com/osm-search/Nominatim/blob/00b3ace3cf92978253557290b0ef12e85d05d7bf/src/nominatim_api/results.py#L460) the appropriate function to compute the display name according to whatever format was chosen in the API parameter

Ideally this works in a way that people can plug in arbitrary functions by putting formatting modules in their project directory. But in a first version we can go with having simply two hard-coded format functions: one for the display name as is now and one for the Open-Cage addresser formatted one.

**Comment by otbutz** [comment]
A really naive question, but is there a 1:1 relationship between the country of an address and its format?

**Comment by lonvia** [comment]
Yes and no. Countries do usually have an official format. So yes. But in pretty much every country you can find addresses that don't fit into the official structure. With those we don't do well with in OSM either. There also is more often than not a divide between rural and city addressing.

The other interesting question is if we'd really always want to use the local format for display, especially given that there are left-to-right and right-to-left readings. But that's to decide once, we have an actual implementation. (But it is the reason why I would want to keep the current display_name formatting as a choice, possibly even the default one.)

**Comment by simon04** [comment]
> simply two hard-coded format functions: one for the display name as is now and one for the Open-Cage addresser formatted one

https://github.com/OpenCageData/address-formatting links to the Python implementation https://github.com/pudo/addressformatting/tree/master which in turn is an archived repository and refers to https://github.com/opensanctions/rigour/blob/main/rigour/addresses/format.py which in turn uses jinja2 to build a formatter for the templates specified in the YAML file https://github.com/OpenCageData/address-formatting/blob/master/conf/countries/worldwide.yaml

tl;dr use Python's yaml and jinja2 to parse the templates in https://github.com/OpenCageData/address-formatting/blob/master/conf/countries/worldwide.yaml
